docs/winnt/cross-build-for-mingw64.md

```diff
@@ -0,0 +1,173 @@
+
+## Cross-compiling NCurses Library
+
+The only outside dependency we need is [ncurses](https://invisible-island.net/ncurses/). Use the latest stable release. Support specifically targeted for MinGW-based ports was added in 5.8. These provide all the features of a normal terminal inside of standard Windows Console .The latest stable release is best here (6.3 as of writing).
+
+There is a port of ncurses called PDCurses that was designed to specifically support DOS and Windows, but it missing a sizable portion of widely used features. The general consensus is not use it if possible.
+
+Grab the source for ncurses, [*https://ftp.gnu.org/pub/gnu/ncurses/ncurses-6.3.tar.gz*](https://ftp.gnu.org/pub/gnu/ncurses/ncurses-6.3.tar.gz). Extract into a directory of your choosing, change into the that directory. Configure with these options:
+
+```
+./configure \
+	--host=$HOST_PLATFORM \
+	--prefix=$MINGW_ROOT \
+	--disable-getcap \
+	--disable-hard-tabs \
+	--disable-home-terminfo \
+	--disable-leaks \
+	--disable-macros \
+	--disable-overwrite \
+	--disable-shared
+	--disable-termcap \
+	--enable-assertions \
+	--enable-database \
+	--enable-echo   \
+	--enable-exp-win32 \
+	--enable-ext-funcs \
+	--enable-interop \
+	--enable-opaque-curses \
+	--enable-opaque-form \
+	--enable-opaque-menu \
+	--enable-opaque-panel \
+	--enable-pc-files \
+	--enable-sp-funcs \
+	--enable-static\
+	--enable-term-driver \
+	--enable-warnings \
+	--enable-widec \
+	--with-fallbacks=ms-terminal \
+	--with-normal \
+	--with-progs \
+	--without-ada \
+	--without-debug \
+	--without-libtool \
+	--without-manpages \
+	--without-shared \
+```
+This configuration is based on the recommended ones for compilation to Windows from NCurses developers ( [README.MinGW](https://github.com/mirror/ncurses/blob/master/README.MinGW) with the addition of `--enable-widec` to provide Unicode support. Script `build_ncurses.sh` is provided to configure with these options.
+
+Standard `make` and `make install` to build the library and install it and its headers into `$MINGW_ROOT`
+
+## Configure and Cross-compile htop
+
+This should be everything we need to cross compile `htop` itself.
+
+Clone the `htop` repo and checkout the `windows` branch (currently work on the `winnt` Port is separated into that branch):
+
+```
+    git clone https://github.com/htop-dev/htop.git
+    cd htop
+    git checkout windows
+```
+
+To configure `htop` for cross-compiling we just need to tell autoconf we want to target the desired platform, and make sure that it looks to our `$MINGW_ROOT` for ncurses. Script `build_htop.sh` is provided to configure with these options.
+
+```
+./autogen.sh
+
+# Make sure configure can find the cross0built ncurses library and headers
+
+export CFLAGS="-I$MINGW_ROOT/include"
+export LDFLAGS="-L$MINGW_ROOT/lib"
+
+./configure \
+	--host=$HOST_PLATFORM \
+	--prefix=$MINGW_ROOT \
+	--enable-static \
+	--enable-unicode
```

Comment:
This fails to find `sys/utsname.h`, as that one is listed in `configure.ac` to be required. Instead when targeting platform `winnt` this should be skipped in `configure.ac`.

Workaround:

```diff
diff --git a/configure.ac b/configure.ac
index 53163366..9893f500 100644
--- a/configure.ac
+++ b/configure.ac
@@ -138,7 +138,6 @@ AC_CHECK_HEADERS([ \
     strings.h \
     sys/param.h \
     sys/time.h \
-    sys/utsname.h \
     unistd.h
    ], [], [AC_MSG_ERROR([can not find required generic header files])])
 
@@ -160,6 +159,10 @@ if test "$my_htop_platform" = darwin; then
    AC_CHECK_HEADERS([mach/mach_time.h])
 fi
 
+if test "$my_htop_platform" != unsupported; then
+   AC_CHECK_HEADERS([sys/utsname.h], [], [AC_MSG_ERROR([can not find required generic header files])])
+fi
+
 # ----------------------------------------------------------------------
 
 
```

Ideally this should be handled as a separate `winnt` platform, but this is upcoming work.

Reply:
> Workaround:

FYI gnulib has a stub for this, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

Reply:
> > Workaround:
> 
> FYI gnulib has a stub header and implementation for this on MinGW/MSVC, See: https://www.gnu.org/software/gnulib/manual/html_node/sys_002futsname_002eh.html and https://github.com/coreutils/gnulib/blob/f36daeac96c30aa5ed42ffa9ccd8cd4366b08b13/lib/uname.c#L56

The `winnt` build shouldn't be using `utsname.h` (it's functionality isn't available on Windows). We want its contents to be missing so that builds that include break ASAP.